Simplify implementation of YAML witnesses

schuessf opened the pull request:
This PR attempts to simplify the implementation for YAML witnesses. The format has completely changed just before SV-COMP 2024 and thus the previous implementation for YAML witnesses was just an ad-hoc adaptation to the changed format.

* Firstly the format for function contracts was changed, such that function contracts are also contained in the `invariant_set` (as in version 2.0), therefore the format for function contracts is downwards compatible and thus the format version was changed from 3.0 to 2.1
* Previously each `WitnessEntry` had the methods `toMap` and `toSetEntry` for the string-conversion to format version 0.1 and 2.0 respectively. Instead of these methods, we have now two different writer-classes that decide which entries should be considered (function contracts should only be exported in version >= 2.1) and how the entries should be formatted in YAML.
* The witness entry classes were initially made to represent the AST. However, since there are now multiple format versions, we use the classes just as container objects to store the necessary informations (e.g. no metadata). Therefore several classes were removed and we avoid parsing the unnecessary informations in `YamlWitnessParser`. Metadata are only needed for witness export, which are now produced using the introduced class `MetadataProvider`
* Previously each witness entry had a set of labels as an unique identifier, to ensure in `MainDispatcher` that each entry is only instrumented once. This is however only specific to GraphML-witnesses. Therefore I added an interface `IExtractedCorrectnessWitness` as an abstraction for the extracted witness entries from GraphML and YAML witnesses and only the GraphML-implementation uses these labels internally. For YAML witnesses we can already ensure that each entry is only instrumented once during extraction.

After this PR is merged, #653 needs to be reworked.

schuessf committed:
Refactor extraction of YAML correctness witnesses
We always generate the same witness independent from the format version.
Therefore WitnessSetEntry was removed and we have different writers instead.
The toMap method was removed from WitnessEntry, since it was specific to format version 0.1.
Only some common classes for the different format versions still provide a toMap method.

schuessf committed:
Refactor witness objects
- only parse the entries from the YAML witness that are necessary for the validation
- produce the other entries only in YamlWitnessWriter without storing them in the witness object itself
- remove Invariant, Metadata, Producer, Task and produce nested maps directly in MetadataProvider and YamlWitnessWriter

schuessf committed:
Remove label from witness entries
This label is only required for GraphML. Instead there is now an interface IExtractedCorrectnessWitness with two implementations, one for GraphML and one for YAML, and only the GraphML implementation uses the label internally (using the internal class LabeledInvariant).
Therefore it is no more required to parse the UUID from the YAML witness.

schuessf committed:
Do not check for FormatVersion
This should be done by witness linter in SV-COMP

schuessf committed:
Match entries from YAML witness at most once

danieldietsch commented on trunk/examples/witness-checking/regression/correctness/contract.c-witness.yml:
```diff
@@ -15,13 +15,13 @@
+  - function_contract:
+      location:
+        file_name: contract.c
+        file_hash: UNUSED
```
😅

danieldietsch commented on trunk/examples/witness-checking/regression/correctness/contract.c-witness.yml:
```diff
@@ -1,8 +1,8 @@
-    format_version: '3.0'
-    uuid: 45befebe-7a0b-4bac-9d4f-25b33879955f
-    creation_time: '2024-02-06T11:15:01+01:00'
+    format_version: '2.1'
```
Why is the format version downgraded?

schuessf replied:
There was never an official version 3.0. We wanted to extend version 2.0 to also support function contracts. Our initial idea was to completely change the format, therefore we also wanted to go from version 2.0 to 3.0. Recently we decided to change the format such that it is downwards compatible (you can see it also in this witness file), therefore we also changed the format version to 2.1.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/GraphMLExtractedCorrectnessWitness.java:
```diff
@@ -11,11 +38,17 @@
 	private final Set<ImmutableSet<String>> mNodeLabelsOfAddedWitnesses = new HashSet<>();
 
-	public GraphMLExtractedCorrectnessWitness(
+	GraphMLExtractedCorrectnessWitness(
```
Do you omit the access modifier to want the constructor and methods to be package-private?
If you don't want to use them outside of the package in the future, then that's a good decision.

schuessf replied:
Yes, the constructor should be only accessible during construction by `GraphMLCorrectnessWitnessExtractor`, only the getters should be public. So maybe it would be even better to make this implementation a nested class in `GraphMLCorrectnessWitnessExtractor` (as it is also rather small and simple).

bahnwaerter replied:
Sure, you could implement it as a nested class, but I think you can leave it as it is.

bahnwaerter commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/YamlExtractedCorrectnessWitness.java:
```diff
@@ -47,15 +47,15 @@ public class YamlExtractedCorrectnessWitness implements IExtractedCorrectnessWit
 	}
 
-	public void addFunctionContract(final IASTNode function, final ExtractedFunctionContract contract) {
+	void addFunctionContract(final IASTNode function, final ExtractedFunctionContract contract) {
```
Do you omit the access modifier to want the methods to be package-private?
If you don't want to use them outside of the package in the future, then that's a good decision.

schuessf replied:
Yes, these methods should be only accessible during construction by `YamlCorrectnessWitnessExtractor`, only the getters should be public. So maybe it would be even better to make this implementation a nested class in `YamlCorrectnessWitnessExtractor` (as it is also rather small and simple).

bahnwaerter replied:
Sure, you could do it as a nested class, but I think you can leave it as it is.

bahnwaerter reviewed (commented):
Your changes look great. Especially, the documentation comments increase code quality.
If omitting the access modifier is intentional, then I have nothing more to complain about.

danieldietsch commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/GraphMLCorrectnessWitnessExtractor.java:
```diff
@@ -178,17 +180,21 @@ private static ExtractedWitnessInvariant mergeInvariants(final ExtractedWitnessI
 			newInvariant = builder.toString();
```
why not just return ewi1?

schuessf replied:
Here we create the disjunction of the invariants from ewi1 and ewi2, also with the union of their node labels.

danieldietsch commented on trunk/source/WitnessParser/src/de/uni_freiburg/informatik/ultimate/witnessparser/YamlWitnessParser.java:
```diff
@@ -80,112 +71,51 @@ public static Witness parseWitness(final File yamlInput) throws IOException {
-			final Invariant locationInvariant = parseInvariant((Map<String, String>) entry.get(LocationInvariant.NAME));
-			return Stream.of(new LocationInvariant(metadata, location, locationInvariant));
+			final Map<String, String> invariant = (Map<String, String>) entry.get(LocationInvariant.NAME);
+			return Stream.of(new LocationInvariant(location, invariant.get("string"), invariant.get("format")));
```
We got rid of the ugly cast, but now we have hardcoded field names in here?

danieldietsch replied:
Ah, the cast is there as well ...

schuessf replied:
This PR did not really change the behavior, the cast and the hardcoded field names already have been there before. With this PR we only extract some necessary fields and without the usage of another method. But I agree that both of the issues are not nice, for the cast I cannot see another solution...

danieldietsch replied:
Yeah, that would require even more work

danieldietsch commented on trunk/source/WitnessPrinter/META-INF/MANIFEST.MF:
```diff
@@ -12,7 +12,8 @@ Require-Bundle: de.uni_freiburg.informatik.ultimate.lib.core,
  de.uni_freiburg.informatik.ultimate.plugins.generator.rcfgbuilder,
- de.uni_freiburg.informatik.ultimate.witnessparser
+ de.uni_freiburg.informatik.ultimate.witnessparser,
+ org.yaml.snakeyaml;bundle-version="2.2.0"
```
Why this version constraint?

schuessf replied:
I am not sure, I simply copied this line from the manifest of `WitnessParser`. Is it better to ommit this version constraint?

danieldietsch replied:
It is definitely better. In particular if there is only one snakeyml.

danieldietsch commented on trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/preferences/PreferenceInitializer.java:
```diff
@@ -105,7 +105,7 @@ public enum WitnessVerifierType {
 
 	public static final String LABEL_YAML_FORMAT_VERSION = "YAML Format version";
-	private static final String VALUE_YAML_FORMAT_VERSION = "3.0";
+	private static final String VALUE_YAML_FORMAT_VERSION = "2.0";
```
This version downgrade really confuses me. 
Furthermore, it looks like we dont have any way to support different versions, right? Do we even check whether the format version is the one we actually support?

schuessf replied:
For the SV-COMP version 2.0 should be still the standard. Therefore I decided to change the default from the experimental version (previously 3.0, now 2.1) to the "official" version 2.0.

The format version is checked in `YamlWitnessWriter::construct`, where we try to get the writer that matches the format version. If we do not support the format version, we throw an exception.

danieldietsch replied:
So basically, having 3.0 was the actual mistake.

schuessf replied:
Yes, if we want to have a "stable" version as a default 😉

danieldietsch commented on trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/yaml/YamlCorrectnessWitnessGenerator.java:
```diff
@@ -40,72 +34,62 @@ public class YamlCorrectnessWitnessGenerator {
-		return new Witness(List.of(witness.toInvariantSet(metadataSupplier)));
+		final var invariants = extractInvariants(allProgramPoints);
+		final var functionContracts = extractFunctionContracts(allProgramPoints);
+		return new Witness(DataStructureUtils.concat(invariants, functionContracts));
```
Is the sorting taken care of now? Or does this not apply anymore?

schuessf replied:
This TODO got lost during refactoring 😅
If we do not sort, we rely on the order of `mIcfg.getProgramPoints()`, which seems to be consistent for `BasicIcfg` the very least, so maybe this is not a critical issue -- it might however not hurt to keep the TODO.

danieldietsch commented on trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/yaml/YamlWitnessWriter.java:
```diff
@@ -0,0 +1,34 @@
+		if (formatVersion.getMajor() == 2) {
+			return new YamlWitnnesWriterV2(metadataProvider, formatVersion.getMinor() >= 1);
+		}
+		throw new UnsupportedOperationException("Unknown format version " + formatVersion);
```
I am getting even more confused about versions now. Here we seem to talk about 0 and 2, but throw an exception for 3?

schuessf replied:
There was the format version 0.1 before the SV-COMP24, in SV-COMP itself version 2.0 was used and is still the current version (version 1.x was simply skipped 😅). We plan to extend this version to be downwards compatible (this already is implemented for function contracts), therefore we use version 2.1 for these extensions. There is currently no version 3.0 (not even "unofficial").

danieldietsch commented on trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/yaml/YamlWitnessWriterV0.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	public String toString(final Witness witness) {
+		return formatYaml(
```
Why are all other entry types thrown away?

schuessf replied:
We now always create the same `Witness`-object, independant from the format version. Therefore this object may also contain e.g. function contracts that are not supported in version 0.1. The writer now decides which entries should be exported to YAML.

schuessf committed:
Change format for function_contract
- function_contracts are contained in the invariant_set
- the format for location_invariant and loop_invariant remains unchanged
- since the format is now downwards compatible, the version number changes from 3.0 to 2.1
- see https://gitlab.com/sosy-lab/benchmarking/sv-witnesses/-/tree/2.1-function-contracts-dev for the format for function contracts

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedFunctionContract.java:
```diff
@@ -77,9 +76,15 @@ private ACSLNode parseString(final ILocation loc, final String string) {
 	}
 
-	public Stream<ACSLNode> transformToAcsl() {
+	public List<ACSLNode> getAcslContracts() {
```
_minor:_ should this be `getAcslContractClauses` or similar? I was confused by the idea that there would be multiple contracts.

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/GraphMLCorrectnessWitnessExtractor.java:
```diff
@@ -105,35 +106,34 @@ protected ExtractedCorrectnessWitness extractWitness() {
-	protected Map<IASTNode, ExtractedWitnessInvariant> mergeMatchesIfNecessary(
-			final Map<IASTNode, ExtractedWitnessInvariant> mapA, final Map<IASTNode, ExtractedWitnessInvariant> mapB) {
+	protected Map<IASTNode, Pair<ImmutableSet<String>, ExtractedWitnessInvariant>> mergeMatchesIfNecessary(
+			final Map<IASTNode, Pair<ImmutableSet<String>, ExtractedWitnessInvariant>> mapA,
```
Would it make sense to introduce a dedicated data type for `Pair<ImmutableSet<String>, ExtractedWitnessInvariant>` ?

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/YamlCorrectnessWitnessExtractor.java:
```diff
@@ -197,11 +195,46 @@ protected int genericVisit(final IASTNode node) {
-				// skip the subtree if a match occurred, but continue with siblings.
-				return PROCESS_SKIP;
+				mMatchedNode = node;
+				// Stop the subtree, since we want the witness entry to be matched at most once.
```
"skip the subtree" :stuck_out_tongue:

maul-esel reviewed (approved):
This looks good to me. Thanks for all the work on improving this stuff :)
